tests/pytests/unit/states/test_pkg.py

```diff
@@ -0,0 +1,244 @@
+        ("< 15.0.0", [("<", "15.0.0")]),
+        ("<15.0.0", [("<", "15.0.0")]),
+        ("15.0.0", [("==", "15.0.0")]),
+        ("", []),
+    ],
+)
+def test_parse_version_string(version_string, expected_version_conditions):
+    version_conditions = pkg._parse_version_string(version_string)
+    assert len(expected_version_conditions) == len(version_conditions)
+    for expected_version_condition, version_condition in zip(
+        expected_version_conditions, version_conditions
+    ):
+        assert expected_version_condition[0] == version_condition[0]
+        assert expected_version_condition[1] == version_condition[1]
+
+
+@pytest.mark.parametrize(
+    "version_string, installed_versions, expected_result",
+    [
+        ("> 1.0.0, < 15.0.0, != 14.0.1", [], False),
+        ("> 1.0.0, < 15.0.0, != 14.0.1", ["1.0.0"], False),
+        ("> 1.0.0, < 15.0.0, != 14.0.1", ["14.0.1"], False),
+        ("> 1.0.0, < 15.0.0, != 14.0.1", ["16.0.0"], False),
+        ("> 1.0.0, < 15.0.0, != 14.0.1", ["2.0.0"], True),
+        ("> 1.0.0, < 15.0.0, != 14.0.1", ["1.0.0", "14.0.1", "16.0.0", "2.0.0"], True,),
+        ("> 15.0.0", [], False),
+        ("> 15.0.0", ["1.0.0"], False),
+        ("> 15.0.0", ["16.0.0"], True),
+        ("15.0.0", [], False),
+        ("15.0.0", ["15.0.0"], True),
+        # No version specified, whatever version installed. This is threated like ANY version installed fulfills.
+        ("", ["15.0.0"], True),
+        # No version specified, no version installed.
+        ("", [], False),
+    ],
+)
+def test_fulfills_version_string(version_string, installed_versions, expected_result):
+    msg = "version_string: {}, installed_versions: {}, expected_result: {}".format(
+        version_string, installed_versions, expected_result
+    )
+    assert expected_result == pkg._fulfills_version_string(
+        installed_versions, version_string
+    ), msg
+
+
+@pytest.mark.parametrize(
+    "installed_versions, operator, version, expected_result",
+    [
+        (["1.0.0", "14.0.1", "16.0.0", "2.0.0"], "==", "1.0.0", True),
+        (["1.0.0", "14.0.1", "16.0.0", "2.0.0"], ">=", "1.0.0", True),
+        (["1.0.0", "14.0.1", "16.0.0", "2.0.0"], ">", "1.0.0", True),
+        (["1.0.0", "14.0.1", "16.0.0", "2.0.0"], "<", "2.0.0", True),
+        (["1.0.0", "14.0.1", "16.0.0", "2.0.0"], "<=", "2.0.0", True),
+        (["1.0.0", "14.0.1", "16.0.0", "2.0.0"], "!=", "1.0.0", True),
+        (["1.0.0", "14.0.1", "16.0.0", "2.0.0"], "==", "17.0.0", False),
+        (["1.0.0"], "!=", "1.0.0", False),
+        ([], "==", "17.0.0", False),
+    ],
+)
+def test_fulfills_version_spec(installed_versions, operator, version, expected_result):
+    msg = "installed_versions: {}, operator: {}, version: {}, expected_result: {}".format(
+        installed_versions, operator, version, expected_result
+    )
+    assert expected_result == pkg._fulfills_version_spec(
+        installed_versions, operator, version
+    ), msg
```

Comment:
```suggestion
    assert expected_result == pkg._fulfills_version_spec(
        installed_versions, operator, version
    )
```
We don't really need to pass that `, msg` since Pytest does an awesome job on giving context when it fails.

Reply:
the same for the other asserts